lib/faraday/request/multipart_with_file.rb

```diff
@@ -5,10 +5,26 @@ module Faraday
   # @private
   class Request::MultipartWithFile < Faraday::Middleware
     def call(env)
+
       if env[:body].is_a?(Hash)
+
+        # Check for IO (and IO-like objects, like Zip::InputStream) in the request,
+        # which represent data to be uploaded.  Replace these with Faraday
         env[:body].each do |key, value|
-          if value.is_a?(File)
-            env[:body][key] = Faraday::UploadIO.new(value, mime_type(value), value.path)
+
+          # Faraday seems to expect a few IO methods to be available, but that's all:
+          # https://github.com/lostisland/faraday/blob/master/lib/faraday/file_part.rb
+          # :length seems to be an optional one
+          #
+          # UploadIO also seems to do a duck typing check for :read, with :path optional
+          # https://www.rubydoc.info/gems/multipart-post/2.0.0/UploadIO:initialize
+          #
+          # We attempt to make our duck typing compatible with their duck typing
+          if value.respond_to?(:read) && value.respond_to?(:rewind) && value.respond_to?(:close)
+            # Mimic UploadIO's handling of inputs like StringIO, which don't have a :path method
+            local_path = value.respond_to?(:path) ? value.path : 'local.path'
+
+            env[:body][key] = Faraday::UploadIO.new(value, mime_type(value), local_path)
```

Comment:
This seems to be repeating the filename/local_path logic in UploadIO initializer here. Am I missing something again?

https://github.com/socketry/multipart-post/blob/bfadd4f3f27c41d4661bca0a64b254b7ef83f426/lib/multipart/post/composite_read_io.rb#L88

```suggestion
            env[:body][key] = Faraday::UploadIO.new(value, mime_type(value))
```

Reply:
Aah, I think I misread the logic in `initialize`, it looks like line 79 ensures that `filename` will never be `nil` (a case that I thought I had to explicitly avoid).  Yes, your fix is the correct one.